README
		
		
		WILLIAMS SYSTEM EDITOR
		Jerry Lee Williams Jr
		25 May 2010
		
		

KEY REFERENCE
	^A	Select all
	^B	Match brace
	^C	Copy
	^D	Duplicate line
	^E	
	^F	Find...
	^_F	Find next
	^G	Go to line...
	^H	Backspace
	^I	Tab (indent if there's a selection)
	^_I	Tab (unindent if there's a selection)
	^J	Join (append next line to this)
	^K	
	^L	Delete this line
	^M	Linebreak (Enter)
	^_M	Insert line below (^Enter)
	^N	New file
	^O	Open file
	^P	Insert line above
	^_P	Insert line above and below
	^Q	Comment/Uncomment selection (or current line)
	^R	Find & Replace...
	^S	Save
	^T	
	^U	
	^V	Paste
	^W	
	^X	Cut
	^Y	Redo
	^Z	Undo
	
	^Space	Select current word
	^Bksp	Delete previous word
	^Del	Delete next word
	_Del	Cut
	^Ins	Copy
	_Ins	Paste
	Home	Go to the first character of the line
	^Home	Go to the irst character of the file
	End	Go to the Last character of the line
	^End	Go to the Last character of the file
	^Left	Previous word
	^Right	Next word
	^Up	Move line up
	^Down	Move line down
	PgUp	Move up one page
	PgDown	Move down one page
	
	F2	Open another editor
	^F2	Spawn a shell
	F3	Find Next
	_F3	Find Next in opposite direction
	F5	Reload this file
	F7	Spawn previous command
	^F7	Spawn command
	F12	Next configuration
	^F12	Reload configuration file
	_F12	Previous configuration

REGULAR EXPRESSIONS
	Searching is always done via regular expressions. The
	particular kind of expressions used are based on the
	POSIX Basic Regular Expressions (4th Edition of the
	Single Unix Specification). Some additions from the
	Extended Regular Expression and PERL were added and
	some features of the POSIX specification are missing as
	noted below.
	
	ABSENT FEATURES
		Equivalence classes
		Coallating classes
		Character classes
	
	ESCAPES
		Escape sequences are not defined by in BREs.
		This is an extension based on ISO C's. Further,
		they are allowed in bracketed expressions (but
		not in ranges).
		
		\t	tab
		\n	linefeed
		\r	carriage return
		\\	backslash
		\f	form feed
		\v	vertical tab
	
	ANCHORS
		^ and $ always act as anchors and never as
		ordary characters except in bracketed
		expressions. (BRE allows this).
	
	EXTENDED OPERATORS
		These operators were taken from the Extended
		Regular Expression specification.
	  	
		+	one or more repetitions
		?	zero or one repetitions
	
	PERL-STYLE CLASSES
		
		\a	lowercase letter
		\A	uppercase letter
		\b	word boundary anchor [^a-zA-Z0-9_]
		\B	word boundary anchor [a-zA-Z0-9_]
		\d	digit
		\D	non-digit
		\s	space
		\S	non-space
		\w	word character [a-zA-Z0-9_]
		\W	non-word character [a-zA-Z0-9_]
		\<	beginning of word anchor
		\>	end of word anchor
	

changes.txt
22 May 2010
	Join compresses/inserts space for ^J, not for Del/Bksp
	Multi-line join
	Braces match through nesting
	Can select with matching braces now
	^[ changed to ^B because char 27 isn't sent with shift
	Menu changes
	Added shell on ^F2
	Run command from _F2
	Stopped the cursor creep in comment/uncomment
	Added support for selecting language based on extension
	Primitive syntax highlighting for comments
	Lines wider than the screen are clipped
24 May 2010
	Loading a file with no selection and the same number of
	  lines would not snap the caret or invalidate.
	Reloading a file attempts to stay on the same line.
25 May 2010
	Clipping corrected, though it over-estimates with tabs
	Fixed the bug where the cursor would initially appear
	  to be on the second line.
	Regular expression search/replace
	Fixed the bug where the window would be created too low
	Fixed bracketing error bug where you could not jump back
	  over nested brackets.
	Shift+Enter now inserts line below and ^P moves the
	  caret to the line it inserts above.
16 July 2010
	Fixed ^up so that the caret follows the line up.
	Brace selection includes close brace now; and close
	  BEFORE the caret is matched.
18 October 2010
	Adding support for utf-16 input and output.
	Changed share mode to share for read & write.
	Filter control keys that are not mapped.
20 October 2010
	Open... and Save As... start in file path.
	Reload as utf-8 & utf-16.
	Added an Edit to the config menu.
	Spawn commands.
21 October 2010
	Spawned commands are now run through cmd.exe.
	Spawned commands are paused after execution.
	Output of spawned commands can be captured.
31 October 2010
	Fixed bug that added extra linebreak after last line.
07 November 2010
	Fixed bug that closed the window if a command fails.
	Fixed regular expression escapes to include special chars.
	Esc no longer generates a character.
21 November 2010
	Status bar isn't included in InvalidateRect().
	Added doublebuffer option.
	Removed the default configuration.
	No new line is loaded if file doesn't end in LF.
	Switched to 64-bit.
	Added syntax highlighting (kwd).
04 December 2011
	Window no longer resizes on config reload.
	Double-click and ^Space selects a word.
	4 wires now; dotted instead of solid lines.
	cp1252 encoding added.
	Detect and use crlf linebreaks.
	Changed kwd format to use '[b][i]' for bold/italic.
	Added drag & drop.
	Spawning programs does not bring up console.
	Breaking after "a" in "a b" leaves the space before "b".
	Find & Replace boxes is populated with selection.
06 December 2011
	Errors in configuration file are now silently ignored.
	Booleans now use yes|no.
	Lowered the number of colours to 8.
	Allow comments to start in any column in config.
	Changed the names of almost all fields.
	Nothing is done if comment is empty.
	Add odd line colouring.
	Status now shows number of chars selected.
	Encoding menu shows 4/8 EM tab and hard/soft tabs.
09 December 2011
	Added support for BOM in UTF-8 & UTF-16.
	Added BOM and CRLF config options.
	Fixed off-by-one error in line counter on status bar.
10 December 2011
	Fixed error where each line is signed with BOM.
	Set default to UNIX linebreaks and no BOM.
	Cleared language count fixing config reload error.
	Searches directories on reload config.
19 December 2011
    Made the "cmd /c %ls && pause >nul" optional in cmd_wrapper.
    Changed c0-c7 to style0 and added bold/italic settings.
    Removed bold/italic settings from keywords.
    wse.conf is no longer loaded from local directory.
03 January 2012
	Fixed BOM encoding w/ UTF-8.
	Visible line count is updated when font config is changed.
	Fixed syntax highlighting to work with puncutation.
	'true' can be used as a value for booleans.
	.9 weight causes bold style to be .4 weight.
08 January 2012
    Keywords no longer required to end with break.
    Added \b and \~, break and backspace to keyword expressions.
    No tabs & 4-character tabbing is the default now.
    vi modelines: tabstop=*,[no]expandtab.
    Left and right now move over tab spaces as a group.
    Per-file settings: tab_width, use_tab, use_crlf, use_bom.
    Added HSV colours: [0-360]° [0,1] [0,1] (° = Alt+0176).
17 January 2012
    Clipping applied to dirty bitmap sections.
28 January 2012
    Added visual styles to dialog boxes.
    Run command input box automatically scrolls.
25 September 2012
    Added HSY colour space.
    Changed format for HSL colour space.
18 October 2012
    Fixed error where last line cannot be joined.
    Fixed Run... dialog and added run w/o console.
24 March 2012
    Put #pragma comments instead of listing libraries on commandline
    Added blurred rendering

conf.c
/* vim: set noexpandtab:tabstop=8 */
#define _WIN32_WINNT 0x0501
#define WIN32_LEAN_AND_MEAN
#define UNICODE
#include <Windows.h>
#include <math.h>
#include <stdlib.h>
#include <wchar.h>
#include "conf.h"
#include "wse.h"
#include "re.h"

enum {
	Boolean,
	Int,
	Color,
	Style,
	Float,
	String,
	Keyword
};

struct field {
	wchar_t	*name;
	int	type;
	void	*ptr;
};

static struct	field fields[] = {
		{L"bg_color", Color, &conf.bg},
		{L"bg_color2", Color, &conf.bg2},
		{L"fg_color", Color, &conf.fg},
		{L"select_color", Color, &conf.selbg},
		{L"bg_image", String, &conf.bgimage},
		{L"doublebuffer", Boolean, &conf.doublebuffer},
		
		{L"style0", Style, &conf.style[0]},
		{L"style1", Style, &conf.style[1]},
		{L"style2", Style, &conf.style[2]},
		{L"style3", Style, &conf.style[3]},
		{L"style4", Style, &conf.style[4]},
		{L"style5", Style, &conf.style[5]},
		{L"style6", Style, &conf.style[6]},
		{L"style7", Style, &conf.style[7]},
		
		{L"font_name", String, &conf.fontname},
		{L"font_size", Float, &conf.fontsz},
		{L"font_aspect", Float, &conf.fontasp},
		{L"font_weight", Float, &conf.weight},
		{L"font_smoothing", Float, &conf.smooth},
		{L"font_italic", Boolean, &conf.italic},
		{L"line_height", Float, &conf.leading},
		{L"font_blur", Float, &conf.blur},
		{L"font_blur_x", Int, &conf.fbx},
		{L"font_blur_y", Int, &conf.fby},
		
		{L"ext", String, &lang.ext},
		{L"comment", String, &lang.comment},
		{L"comment_color", Int, &lang.commentcol},
		{L"break", String, &lang.brk},
		{L"brace", String, &lang.brace},
		{L"kwd", Keyword, 0},
		{L"cmd_wrapper", String, &lang.cmdwrapper},
		
		{L"wire", Int, conf.wire},
		{L"wire2", Int, conf.wire+1},
		{L"wire3", Int, conf.wire+2},
		{L"wire4", Int, conf.wire+3},
		{L"tab_width", Int, &conf.tabc},
		{L"use_tabs", Boolean, &conf.usetabs},
		{L"use_bom", Boolean, &conf.usebom},
		{L"use_crlf", Boolean, &conf.usecrlf},
		{L"cols", Int, &conf.cols},
		{L"rows", Int, &conf.rows},
		
		{L"alpha", Float, &alpha},
		{L"shell", String, &shell},
		{0}
		};

static LOGFONT	lf = {
		0, 0, 0,0, FW_NORMAL, 0,0,0,
		DEFAULT_CHARSET, OUT_DEFAULT_PRECIS,
		CLIP_DEFAULT_PRECIS, DEFAULT_QUALITY,
		FF_DONTCARE|DEFAULT_PITCH, 0
		};
extern HFONT	font[4];

wchar_t		shell[128]=L"cmd";

configfont() {
	HDC		hdc;
	int		dpi,i;
	TEXTMETRIC	tm;
	
	if (font[0]) {
		DeleteObject(font[0]);
		DeleteObject(font[1]);
		DeleteObject(font[2]);
		DeleteObject(font[3]);
	}
	
	/* Get device resolution */
	hdc = GetDC(0);
	dpi = GetDeviceCaps(hdc, LOGPIXELSY);
	
	/* Create font */
	lf.lfHeight = -conf.fontsz * dpi/72.0;
	lf.lfWidth = fabs(lf.lfHeight) * conf.fontasp;
	lf.lfItalic = conf.italic;
	lf.lfWeight = conf.weight * 1000;
	lf.lfQuality = conf.smooth
		? (conf.smooth<=0.5? ANTIALIASED_QUALITY: CLEARTYPE_QUALITY)
		: NONANTIALIASED_QUALITY;
	wcscpy(lf.lfFaceName, conf.fontname);
	font[0] = CreateFontIndirect(&lf); /* Regular */
	
	lf.lfItalic ^= 1;
	font[2] = CreateFontIndirect(&lf); /* Italic */
	
	lf.lfItalic ^= 1;
	lf.lfWeight = lf.lfWeight==900? 400: 900;
	font[1] = CreateFontIndirect(&lf); /* Bold */
	
	lf.lfItalic ^= 1;
	font[3] = CreateFontIndirect(&lf); /* Bold & Italic */
	
	/* Get metrics */
	SelectObject(hdc, font[0]);
	GetTextMetrics(hdc, & tm);
	GetCharWidth32(hdc, 0, 65535, conf.widths);
	ReleaseDC(0, hdc);
	
	conf.aheight = tm.tmHeight;
	conf.lheight = conf.aheight * conf.leading
		+ tm.tmExternalLeading;
	conf.em = tm.tmAveCharWidth;
	conf.tabw = conf.em * conf.tabc;
	return 1;
}

deflang() {
	*lang.ext=0;
	wcscpy(lang.comment, L"");
	wcscpy(lang.brk, L"~!@#$%^&*()-+={}[]\\|;:'\",.<>/?");
	wcscpy(lang.brace, L"()[]{}''\"\"<>``");
	memset(lang.kwd,0,sizeof lang.kwd);
	wcscpy(lang.cmdwrapper, L"cmd /c %ls & pause >nul");
	lang.nkwd=0;
	lang.commentcol=0;
}

defconfig() {
	memset(conf.style, 0, sizeof conf.style);
	conf.bg = RGB(255,255,255);
	conf.bg2 = RGB(245,245,245);
	conf.fg = RGB(64,64,64);
	conf.selbg = RGB(160,160,192);
	conf.style[0].color = RGB(160,160,192);
	wcscpy(conf.bgimage, L"");
	
	conf.doublebuffer=1;
	
	conf.tabc = 4;
	conf.usetabs = 0;
	conf.usebom = 0;
	conf.usecrlf = 0;
	conf.wire[0] = 64;
	conf.wire[1] = 72;
	conf.wire[2] = 80;
	conf.wire[3] = 128;
	conf.cols = 80;
	conf.rows = 24;
	alpha = .9;
	
	wcscpy(conf.fontname, L"Courier New");
	conf.fontsz = 12.0;
	conf.fontasp = 0.0;
	conf.leading = 1.5;
	conf.smooth = 1.0;
	conf.italic = 0;
	conf.weight = .4;
	conf.blur = .2;
	conf.fbx = 0;
	conf.fby = -1;
	
	configfont();
	return 1;
}

static
directive(wchar_t *s) {
	int	x;
	wchar_t	*e;

	while (iswspace(*s))
		s++;
	
	if (!*s) {
		confset[nconfs++]=conf;
		defconfig();
	}
	else if (isdigit(*s))
		selectconfig(wcstol(s, 0, 10));
	else if (*s==L'.') {
		x=0;
		s=lang.ext;
		while (iswspace(*s))
			s++;
		
		while (e=wcschr(s, L',')) {
			*e=0;
			s=e+1;
		}
		
		langset[nlangs++]=lang;
		deflang();
	}
	return 1;
}

getcolor(wchar_t *arg) {
	double		h, s, v; /* hue,chroma,luma */
	double		y; /* luma (Y'601) */
	int		r,g,b;
	if (3 == swscanf(arg, L"hsy %lf %lf %lf", &h,&s,&y)) {
		float x, r1,g1,b1, m;
		h /= 60.0;
		x = s * (1 - fabs(fmod(h, 2) - 1));
		if (0 <= h && h <= 1)	r1 = s, g1 = x, b1 = 0;
		else if (h <= 2)	r1 = x, g1 = s, b1 = 0;
		else if (h <= 3) 	r1 = 0, g1 = s, b1 = x;
		else if (h <= 4) 	r1 = 0, g1 = x, b1 = s;
		else if (h <= 5) 	r1 = x, g1 = 0, b1 = s;
		else if (h <= 6) 	r1 = s, g1 = 0, b1 = x;
		else	 		r1 = 0, g1 = 0, b1 = 0;
		m = y - (.3*r1 + .59*g1 + .11*b1);
		r = 255 * (r1 + m);
		g = 255 * (g1 + m);
		b = 255 * (b1 + m);
		if (r < 0) r = 0; else if (r > 255) r = 255;
		if (g < 0) g = 0; else if (g > 255) g = 255;
		if (b < 0) b = 0; else if (b > 255) b = 255;
		return ((unsigned)b<<16)+
			((unsigned)g<<8)+
			((unsigned)r);
	} else if (3 == swscanf(arg, L"hsl %lf %lf %lf", &h,&s,&v)) {
		double r,g,b, f, p,q,t;
		h /= 60.0;
		f = h - floor(h);
		p = v * (1.0 - s);
		q = v * (1.0 - s * f);
		t = v * (1.0 - s * (1.0 - f));
		if (h < 1.0) r=v, g=t, b=p;
		else if (h < 2.0) r=q, g=v, b=p;
		else if (h < 3.0) r=p, g=v, b=t;
		else if (h < 4.0) r=p, g=q, b=v;
		else if (h < 5.0) r=t, g=p, b=v;
		else r=v, g=p, b=q;
		return ((unsigned)(b*255)<<16)+
			((unsigned)(g*255)<<8)+
			((unsigned)(r*255));
	} else if (3 == swscanf(arg, L"%d %d %d", &r,&g,&b)
	  || 3 == swscanf(arg, L"rgb %d %d %d", &r,&g,&b)) {
		return (b<<16)+(g<<8)+r;
	}
	return 0;
}

static
configline(int ln, wchar_t *s) {
	wchar_t		*arg;
	struct textstyle *style;
	struct field	*cf;
	
	while (iswspace(*s))
		s++;
	if (*s=='#' || !*s)
		return 0;
	if (*s=='.')
		return directive(s+1);
	
	/* Cut field name */
	for (arg=s; *arg && !iswspace(*arg); arg++);
	*arg++=0;
	while (iswspace(*arg)) arg++;
	
	/* Find field */
	for (cf=fields; cf->name && wcscmp(cf->name, s); cf++);
	if (!cf->name)
		return 0;
	
	/* Remove comment if not string or keyword */
	if (cf->type != String && cf->type != Keyword)
		arg[wcscspn(arg, L"#")] = 0;
	
	switch (cf->type) {
	
	case Boolean:
		if (!wcscmp(arg,L"yes")) *(int*)cf->ptr = 1;
		else if (!wcscmp(arg,L"true")) *(int*)cf->ptr = 1;
		else *(int*)cf->ptr = 0;
		while (*++arg);
		return 1;
	
	case Int:
		*(int*)cf->ptr = wcstol(arg, &arg, 0);
		return 1;

	case Color:
		*(int*)cf->ptr = getcolor(arg);
		return 1;
	
	case Style:
		style = (struct textstyle*)cf->ptr;
		style->style = 0;
		for (;;) {
			while (iswspace(*arg))
				arg++;
			if (!wcsncmp(arg, L"bold", 4))
				arg+=4, style->style |= 1;
			else if (!wcsncmp(arg, L"italics", 7))
				arg+=7, style->style |= 2;
			else if (!wcsncmp(arg, L"italic", 6))
				arg+=6, style->style |= 2;
			else
				break;
		}
		style->color = getcolor(arg);
		return 1;
	
	case Float:
		*(double*)cf->ptr = wcstod(arg, &arg);
		return 1;
	
	case String:
		wcscpy(cf->ptr, arg);
		return 1;
	
	case Keyword:
		lang.kwdcol[lang.nkwd] = wcstoul(arg,&arg,0);
		while (iswspace(*arg))
			arg++;
		re_comp(lang.kwd[lang.nkwd], arg);
		lang.nkwd++;
		return 1;
	}
	return 0;
}

loadconfig(wchar_t *fn) {
	HANDLE		file;
	DWORD		sz, ign;
	int		c,ln,eol;
	char		*utf;
	wchar_t		*buf, *s;
	
	file=CreateFile(fn, GENERIC_READ, FILE_SHARE_READ, 0,
		OPEN_EXISTING, 0, 0);
	if (file==INVALID_HANDLE_VALUE)
		return 0;
	
	sz=GetFileSize(file, 0);
	utf=malloc(sz+1);
	ReadFile(file, utf, sz, &ign, 0);
	CloseHandle(file);
	utf[sz]=0;
	
	sz=MultiByteToWideChar(CP_UTF8, 0, utf, sz+1, 0, 0);
	buf=malloc((sz+1) * sizeof (wchar_t));
	MultiByteToWideChar(CP_UTF8, 0, utf, sz+1, buf, sz+1);
	free(utf);
	
	defconfig();
	deflang();
	confset[nconfs]=conf;
	
	s=buf;
	ln=0;
	while (*s) {
		ln++;
		eol=wcscspn(s, L"\r\n");
		c=s[eol];
		if (c=='\r')
			s[eol++]=0;
		else if (c=='\n')
			s[eol]=0;
		
		configline(ln, s);
		
		s += eol + (c!=0);
	}
	free(buf);
	
	configfont();

	return 1;
}

selectconfig(int x) {
	if (nconfs<=x || x<0)
		return 0;
	curconf=x;
	conf=confset[x];
	return 1;
}

config() {
	wchar_t	path[MAX_PATH];
	
	nconfs=0;
	nlangs=0;
	
	GetModuleFileName(0, path, MAX_PATH);
	wcscpy(wcsrchr(path, L'\\')+1, L"wse.conf");
	if (loadconfig(path)) {
		configfile=wcsdup(path);
		return 1;
	}
	configfile=L"";
	defconfig();
	deflang();
	confset[nconfs++]=conf;
	return 0;
}
